04 - Functional Programming/04 - Functional Programming.ipynb

```diff
@@ -0,0 +1,3551 @@
+            "Enter a number: 3\n",
+            "2 + 3 = 5\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Декоратори"
+      ],
+      "metadata": {
+        "id": "V0akTWPMDDTj"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Една от по-интересните функционалности на Python са т.нар. декоратори (decorators). Всъщност, ние вече видяхме как се пише декоратор и как работи - нека се върнем на примера от по-горе с `log_and_run`:"
+      ],
+      "metadata": {
+        "id": "FwuzabXI8nbD"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "def log_and_run(f, args=(), kwargs={}):\n",
+        "    print(f'Calling {f} with {args} and {kwargs}')\n",
+        "    return f(*args, **kwargs)\n",
+        "\n",
+        "def add(a, b):\n",
+        "    return a + b\n",
+        "\n",
+        "print(log_and_run(add, (2, 3)))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "dDgfp7RE9YTQ",
+        "outputId": "64afde38-b780-4d92-fe96-566839557943"
+      },
+      "execution_count": 9,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "Calling <function add at 0x7f2c9d0b7170> with (2, 3) and {}\n",
+            "5\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Това изглежда леко грозно, да трябва да извикваме `log_and_run` вместо функцията, която всъщност искаме да извикаме. Ако добавим функции за изваждане, умножение и деление, всяко извикване ще трябва да преминава през `log_and_run`. \n",
+        "\n",
+        "А няма ли по-лесен начин с който да кажем - нека преди изпълнението на нашата функция `f1`, да се изпълни функция `f` (Както преди `add` се изпълнява `log_and_run`) ?"
+      ],
+      "metadata": {
+        "id": "YyZvOMcn9eyZ"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Преди да разгледаме как точно се прави декоратор функция в Python, нека всъщност разберем какво искаме да постигнем:\n",
+        "1. Искаме да приемем функция (нека я кръстим `g`), която да изпълним\n",
+        "2. Преди да изпълним `g`, искаме да изпълним някакви други действия (нека те са събрани във функцията `f`\n",
+        "3. Искаме да върнем нова функция (нека да я кръстим `h`), която първо да изпълни `f`, а после `g`.\n",
+        "\n",
+        "Нека първо се фокусираме върху 1 и 2 :"
+      ],
+      "metadata": {
+        "id": "_GRsD2OR-zeB"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "def pseudo_decorator(f, g):\n",
+        "    f()\n",
+        "    g()\n",
+        "\n",
+        "\n",
+        "def print_hello():\n",
+        "    print(\"Hello\")\n",
+        "\n",
+        "def print_bye():\n",
+        "    print(\"bye\")\n",
```

Comment:
nit: My OCD is is triggered that "Hello" is capitalized and "bye" isn't

Reply:
Fixed